controllers/datadogagent/controller_reconcile_v2.go

```diff
@@ -140,20 +143,31 @@ func (r *Reconciler) reconcileInstanceV2(ctx context.Context, logger logr.Logger
 		return r.updateStatusIfNeededV2(logger, instance, newStatus, result, err)
 	}
 
-	// if introspection is enabled, for all providers, reconcile the node agent
-	// if introspection is disabled, reconcile the agent once using the empty provider `LegacyProvider`
-	providersList := map[string]struct{}{kubernetes.LegacyProvider: {}}
-	if r.options.IntrospectionEnabled {
-		providersList, err = r.handleProviders(ctx, instance, newStatus)
-		if err != nil {
-			errs = append(errs, err)
-		}
+	profiles, profilesByNode, err := r.profilesToApply(ctx, logger)
+	if err != nil {
+		return reconcile.Result{}, err
 	}
-	for provider := range providersList {
-		result, err = r.reconcileV2Agent(logger, requiredComponents, features, instance, resourceManagers, newStatus, provider)
-		if utils.ShouldReturn(result, err) {
-			return r.updateStatusIfNeededV2(logger, instance, newStatus, result, err)
+
+	for _, profile := range profiles {
+		// if introspection is enabled, for all providers, reconcile the node agent
+		// if introspection is disabled, reconcile the agent once using the empty provider `LegacyProvider`
+		providersList := map[string]struct{}{kubernetes.LegacyProvider: {}}
+		if r.options.IntrospectionEnabled {
+			providersList, err = r.handleProviders(ctx, instance, newStatus)
+			if err != nil {
+				errs = append(errs, err)
+			}
 		}
+		for provider := range providersList {
+			result, err = r.reconcileV2Agent(logger, requiredComponents, features, instance, resourceManagers, newStatus, provider, &profile)
+			if utils.ShouldReturn(result, err) {
+				return r.updateStatusIfNeededV2(logger, instance, newStatus, result, err)
+			}
+		}
+	}
+
+	if err = r.handleProfiles(ctx, profiles, profilesByNode, instance.Namespace); err != nil {
```

Comment:
This function does node labeling, after DSs are already create. What happens if DS creation is successful but labeling fails (or keeps failing)? Do we reconcile such nodes?

Reply:
The error handling logic follows the same pattern as in other places of the code.

All the steps in `handleProfiles` should be successful. If there's an error in any of the steps, we return an error in the Reconcile function so that the request is retried later.

If it keeps failing, we would have the same problem as in any other part of the code that creates, updates or deletes any Kubernetes objects. There isn't much we can do. Although the only case where I think the node labeling could be failing permanently (and not other operations) is if RBACs are not properly set, but that should be taken care of by our installation charts.